dev/core#3722 Stop infinite reconciliation

herbdool opened the pull request:
Overview
----------------------------------------
Fixes https://lab.civicrm.org/dev/core/-/issues/3722

Before
----------------------------------------

It schedules managed entity reconciliation over and over and over and over and over again.

After
----------------------------------------

It only schedules reconciliation if the definition is from an xml file *and* it is in the db already.

Technical Details
----------------------------------------

New to 5.50.0 we can now export CaseTypes as managed entities from API4. This attempts to fix the assumptions in CaseType which thinks it's coming from an xml file alone.

herbdool committed:
Only reconcile if has an xml file

colemanw commented:
I'm unclear what this reconciliation is even for. Why is it needed? And why is it needed only under that condition?
Some code comments might help.

herbdool commented on CRM/Case/BAO/CaseType.php:
```diff
@@ -60,7 +60,9 @@ public static function add(&$params) {
-      CRM_Core_ManagedEntities::scheduleReconciliation();
+      if (self::isForked($params['id'])) {
+        CRM_Core_ManagedEntities::scheduleReconciliation();
+      }
```
I think it still creates them automatically. It's just an extra check here that it is indeed "forked" which then weeds out the case where it's actually from an mgd.php file.

Though this patch has a different issue. On a reinstall with db backup with an enabled extension that has a CaseType mgd.php file I get this error:

```
In DAO.php line 1413:
                        
  getFieldValue failed
```

I don't know why it doesn't like `isForked()` in `CRM_Case_BAO_CaseType::add()` when it seems to be fine in `CRM_Case_BAO_CaseType::create()`. Perhaps instead it should just check that the xml file exists in a new function.

demeritcowboy replied:
Calling isForked if params['id'] is blank will give that error.

herbdool replied:
I've updated the PR to now only check if it's an XML based definition.

herbdool committed:
New method that checks if it has an xml definition

herbdool committed:
Update CaseType.php

herbdool commented:
@colemanw I'm not sure if we can remove the reconciliation altogether, but that would be the cleanest option. I can only imagine it's there because of possible "forked" entities? It only gets called if the `definition` parameter is set which in turn would *not* be set if all these conditions apply: editing a case type, is not yet forked, and is not forkable. Or to be clearer reconciliation will take place if Case Type is forked, or new, or is forkable.

herbdool commented:
@colemanw here's where it was added https://github.com/civicrm/civicrm-core/commit/9c19292b47fdfab36542769d40c322e0d463b87c "CRM-14786 - CaseType - Perform reconciliation whenever a case-type definition changes"

herbdool committed:
Update CaseType.php

herbdool commented:
I added some code comments.

herbdool commented:
Looks like my PR isn't backwards-compatible yet.

demeritcowboy commented:
I was wondering what would happen if you call api casetype.create with a definition. Then there's no physical xml file, so it looks like it's skipping the reconciliation. That's how I interpret the test fail.

herbdool committed:
Update CaseType.php

herbdool commented:
Going to see if it's okay with specifying `'version' => 3`.

demeritcowboy commented:
But then what if you call `\Civi\Api4\CaseType::create()->addValue('definition', ...)` ? That would fail to do its thing.

Ideas:
* add a \Civi::$statics in https://github.com/civicrm/civicrm-core/blob/4ade7791ed68b12a230bf4c0cbd72ca688e0deb6/CRM/Core/ManagedEntities.php#L47 so it only does it once, or
* Store a hash of the contents of each reconciled thing in https://github.com/civicrm/civicrm-core/blob/4ade7791ed68b12a230bf4c0cbd72ca688e0deb6/CRM/Core/ManagedEntities.php#L531, or somewhere in that file, and don't reconcile two things with the same hash twice.

totten commented:
* __Comment__: Basing the behavior on v3-vs-v4 makes more sense than the prior draft that used xml-file-presence.

* __GUI__

    >  In the test it's created it with a fixture, just defined in a variable It doesn't seem like quite a real world test of functionality but I guess it's possible.

    FWIW, IIRC, it was the real-world scenario when this API+GUI+test were introduced. The GUI simply called `CaseType.create` API and passed in the `definition`, and the API would do any maintenance updates on ActivityTypes, RelationshipTypes, etc. The GUI was literally just an editor for the `definition` (which was the source of truth).

    The GUI changed somewhere along the line (*it now opens various quickform popups*), so yeah - it probably doesn't rely on that behavior so much.

* __Auto installation__

    With the recent PR that updated `hook_managed` to support `CaseType`@`v4`, I recall being surprised that it worked - and learning that it worked because it relied on autocreating items from the `definition`. There are a few code-paths to autocreation, and this PR may or may not be impact each. I think the main thing is to test and not leave it up to chance.

    I posted a couple related PRs with tests: #23959 and #23961. 

    Unfortunately, while checking for baseline (`master`, pre-PR), I found that it `hook_managed[CaseType]` seems to crash if you use an actual/exported CaseType with a real `definition`: https://gist.github.com/totten/3ec8164c3f8ca9d3c51e768feab946da

herbdool commented:
When I try the equivalent of `\Civi\Api4\CaseType::create()->addValue('definition', ...)` via the API4 GUI it encodes all the XML tags so it gets stored incorrectly. Not sure if that also happens if doing it directly in code.

As for statics, this seems to work:

```
     if (!isset(\Civi::$statics[__CLASS__][$params['name']])) {
        \Civi::$statics[__CLASS__][$params['name']] = 1;
        CRM_Core_ManagedEntities::scheduleReconciliation();
      }
```

Though you suggested putting it in `CRM_Core_ManagedEntities` so I'll try it there as well.

herbdool commented:
This works too:

```
  public function reconcile($modules = NULL) {
    $modules = $modules ? (array) $modules : NULL;
    $declarations = $this->getDeclarations($modules);
    $plan = $this->createPlan($declarations, $modules);
    if (!isset(\Civi::$statics[__CLASS__]['plan'])) {
      \Civi::$statics[__CLASS__]['plan'] = TRUE;
      $this->reconcileEntities($plan);
    }
  }
```

herbdool committed:
Merge branch 'civicrm:master' into core-3722

herbdool committed:
Update CaseType.php

herbdool commented:
Now I'm trying a static. I ended up putting it into `CRM_Case_BAO_CaseType::add()` because I figured here we can ensure it gets called at least once for each case type.

herbdool committed:
Enable checking activity type

herbdool committed:
Update BunnyDance.mgd.php

demeritcowboy commented on CRM/Case/BAO/CaseType.php:
```diff
@@ -60,7 +60,13 @@ public static function add(&$params) {
+      // @see CRM-14786 Reconcilation added to manage whenever a case-type XML definition changes,
+      // in order to add/remove associated activity and relationship types.
+      if (!isset(\Civi::$statics[__CLASS__][$params['name']])) {
+        \Civi::$statics[__CLASS__][$params['name']] = 1;
```
$params['name'] can be blank, e.g. if calling this to update using id instead of name. There is $caseTypeName which should always be set. But I'm wondering if it even needs to be that specific, since (I think) the reconciliation will reconcile whatever needs reconciling and so just needs to be called once. To be honest I'm not sure exactly how it works, and when PHASE_POST_COMMIT occurs.

In reverse, I suppose there's also the situation where, e.g. during upgrade, this might be an issue, because let's say you have something that does casetype.create in 5.52 and something calls reconciliation, then another casetype.create in 5.53 and something calls reconciliation. That second one won't run if running upgrades via cli since the static will stick.

I'm just thinking of some things to test out.

herbdool replied:
Yeah I meant to use `$caseTypeName`.

Based on your second thought it seems like setting it per `$caseTypeName` might be needed. Unless even that could be repeated across upgrades.

herbdool committed:
Update CaseType.php

herbdool commented:
I can't figure out the failure. Says there were no test failures.

demeritcowboy commented:
When that happens you need to hunt in the console log:

```
1) E2E_Shimmy_LifecycleTest::testLifecycleWithSubprocesses
RuntimeException: Command failed (cd '/home/jenkins/bknix-dfl/build/core-23955-56ru3/web/sites/default/files/civicrm/ext/example-mixin'; cv api3 --in=json 'Extension.enable'):
{"error_code":"already exists","tip":"add debug=1 to your API call to have more info about the error","is_error":1,"error_message":"DB Error: already exists"}


/home/jenkins/bknix-dfl/build/core-23955-56ru3/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:161
/home/jenkins/bknix-dfl/build/core-23955-56ru3/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:119
/home/jenkins/bknix-dfl/build/core-23955-56ru3/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:72
/home/jenkins/bknix-dfl/build/core-23955-56ru3/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:43
/home/jenkins/bknix-dfl/extern/phpunit8/phpunit8.phar:671

2) E2E_Shimmy_LifecycleTest::testLifecycleWithLocalFunctions
CRM_Core_Exception: DB Error: already exists

/home/jenkins/bknix-dfl/build/core-23955-56ru3/web/sites/all/modules/civicrm/api/api.php:135
/home/jenkins/bknix-dfl/build/core-23955-56ru3/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:104
/home/jenkins/bknix-dfl/build/core-23955-56ru3/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:68
/home/jenkins/bknix-dfl/build/core-23955-56ru3/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:54
/home/jenkins/bknix-dfl/extern/phpunit8/phpunit8.phar:671
```

herbdool commented:
Running out of steam here. So the last version of this PR that passed as where I restricted it to v3, though not sure if it would pass against master now.

herbdool commented:
For my own project I can just patch Civi via composer until this gets fixed.

I had an xml file up until recently. I had also specified some activity types in mgd.php files. But when I added another activity type in mgd.php and also specified it in the casetype xml file CiviCRM complained that the activity type already existed. It was only when I switched to just using mgd.php that it worked.

So with a patch I can wait a bit - I don't have to worry about all the permutations on my own project.

herbdool commented:
In case it turns out that checking if it's API4 or API3 managed entity is the best way to go, then I just came across `!CRM_Core_BAO_Managed::isApi4ManagedType($item['entity_type'])` in `CRM_Core_ManagedEntities`.

colemanw commented:
> In case it turns out that checking if it's API4 or API3 managed entity is the best way to go, then I just came across `!CRM_Core_BAO_Managed::isApi4ManagedType($item['entity_type'])` in `CRM_Core_ManagedEntities`.

That function checks if an entity is supported by v4, not whether a particular managed record uses v3 or 4. But that is possible to check as well by inspecting a given record. Something like `if (($record['params']['version'] ?? 3) == 4)`

demeritcowboy commented:
Warming might be too strong a word (grin) but it looks like there's very few obvious uses in universe, and we know there's at least one test that covers the v3 use, so go for it!

herbdool committed:
Try check API version again - with new tests

herbdool commented:
Gr... I don't even see the test `civicrm/tests/phpunit/api/v4/Custom/FalseNotEqualsZeroTest.php` which is failing. Does it exist?

demeritcowboy commented:
```
1) E2E_Shimmy_LifecycleTest::testLifecycleWithLocalFunctions
CRM_Core_Exception: Cannot uninstall extension; disable it first: shimmy

/home/jenkins/bknix-dfl/build/core-23955-5hzs5/web/sites/all/modules/civicrm/api/api.php:135
/home/jenkins/bknix-dfl/build/core-23955-5hzs5/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:104
/home/jenkins/bknix-dfl/build/core-23955-5hzs5/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:69
/home/jenkins/bknix-dfl/build/core-23955-5hzs5/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:54


2) E2E_Shimmy_LifecycleTest::testLifecycleWithSubprocesses
ActivityType "Nibble" should be auto enabled. It's missing.
Failed asserting that null matches expected 'Nibble'.

/home/jenkins/bknix-dfl/build/core-23955-5hzs5/web/sites/default/files/civicrm/ext/example-mixin/tests/mixin/ManagedCaseTypeTest.php:30
/home/jenkins/bknix-dfl/build/core-23955-5hzs5/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:96
/home/jenkins/bknix-dfl/build/core-23955-5hzs5/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:73
/home/jenkins/bknix-dfl/build/core-23955-5hzs5/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:43
```

civicrm-builder commented:
Can one of the admins verify this patch?

colemanw commented:
@herbdool thanks for your work on this. I was able to resurrect this PR in the form of #27430 - I used a slightly different approach to prevent recursion (the new-ish ability of managed entities to target a specific module) and that stopped the test from crashing... then I found out that the test itself was broken. That was a fun one.

herbdool commented:
Good to see this gnarly one fixed!
